### PR TITLE
Fix pod logs in case in case empty data from odh-ca-bundle.crt key

### PR DESCRIPTION
**Pull request by atheo89**
Related to: https://issues.redhat.com/browse/RHOAIENG-4165

## Description
This pull request resolves the issue with pod logs by adjusting it to display an informative message instead of raising an error when encountering the `odh-ca-bundle.crt` key, as there might be instances where the data associated with this key is empty.

## How Has This Been Tested?
- Change the deployment on the odh-notebook-controller by using the image created from this PR 
` quay.io/opendatahub/odh-notebook-controller:pr-264`

#### In case of empty odh-ca-bundle.crt
```
{"level":"info","ts":"2024-03-07T15:00:12Z","logger":"controllers.Notebook","msg":"Certificates in 'ca-bundle.crt' are valid and can be mounted","notebook":"emptycert","namespace":"test"}
{"level":"info","ts":"2024-03-07T15:00:12Z","logger":"controllers.Notebook","msg":"odh-ca-bundle.crt data are empty","notebook":"emptycert","namespace":"test"}
{"level":"info","ts":"2024-03-07T15:00:12Z","logger":"controllers.Notebook","msg":"Certificates in 'odh-ca-bundle.crt' are empty, skipping mounting","notebook":"emptycert","namespace":"test"}
```

Mounts only the valid cert 
image

#### In case of wrong cert 

```
{"level":"info","ts":"2024-03-07T15:02:33Z","logger":"controllers.Notebook","msg":"Certificates in 'ca-bundle.crt' are valid and can be mounted","notebook":"wrondcert","namespace":"test"}
{"level":"error","ts":"2024-03-07T15:02:33Z","logger":"controllers.Notebook","msg":"invalid certificate format for key 'odh-ca-bundle.crt' in ConfigMap odh-trusted-ca-bundle","notebook":"wrondcert","namespace":"test","stacktrace":"{..}
{"level":"error","ts":"2024-03-07T15:02:33Z","logger":"controllers.Notebook","msg":"Error validating certificates for odh-ca-bundle.crt. They cannot be mounted","notebook":"wrondcert","namespace":"test","error":"invalid certificate format for key 'odh-ca-bundle.crt' in ConfigMap odh-trusted-ca-bundle","stacktrace":"{...}
```
Mounts only the valid one
image

#### In case of both certs are valid 

```
{"level":"info","ts":"2024-03-07T15:11:28Z","logger":"controllers.Notebook","msg":"Certificates in 'ca-bundle.crt' are valid and can be mounted","notebook":"valid","namespace":"test"}
{"level":"info","ts":"2024-03-07T15:11:28Z","logger":"controllers.Notebook","msg":"Certificates in 'odh-ca-bundle.crt' are valid and can be mounted","notebook":"valid","namespace":"test"}
```
Mounts both cert 
image
## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 1]
```diff
@@ -367,7 +367,8 @@ func certValidator(cm *corev1.ConfigMap, dataKey string, log logr.Logger) error
 	odhCertData, ok := cm.Data[dataKey]
 	if !ok || odhCertData == "" {
 		// Print a warning if odh-ca-bundle.crt data is empty
-		return fmt.Errorf("Warning: %s data is empty", dataKey)
+		log.Info(fmt.Sprintf("Warning: %s data is empty", dataKey))
+		return nil
```
Comment (harshad16) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:371]: This would cause problem, as line 301 and  316 
would be triggered, include something which is not present.

**Review by harshad16 — CHANGES_REQUESTED** [review]
Did we test this in a cluster, i feel this would result in volumemount of empty bit.

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 2]
```diff
@@ -376,9 +377,11 @@ func certValidator(cm *corev1.ConfigMap, dataKey string, log logr.Logger) error
 		// Attempt to parse the certificate
 		_, err := x509.ParseCertificate(odhBlock.Bytes)
 		if err != nil {
-			return fmt.Errorf("error parsing certificate for key '%s' in ConfigMap odh-trusted-ca-bundle: %v", dataKey, err)
+			log.Error(err, fmt.Sprintf("error parsing certificate for key '%s' in ConfigMap odh-trusted-ca-bundle", dataKey))
+			return nil
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:381]: Actually - is this change intended - we don't want to fail in this case?

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 3]
```diff
@@ -376,9 +377,11 @@ func certValidator(cm *corev1.ConfigMap, dataKey string, log logr.Logger) error
...
 		}
 	} else if len(odhCertData) > 0 {
+		log.Error(nil, fmt.Sprintf("invalid certificate format for key '%s' in ConfigMap odh-trusted-ca-bundle", dataKey))
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:384]: Hm, does this mean that we log this error here and then also again where the `certValidator` function is used and checked for the return error (lines 313 and 328)?

`odh-trusted-ca-bundle` - would deserve some constant, but that's just nitpick

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 4]
```diff
@@ -367,7 +367,8 @@ func certValidator(cm *corev1.ConfigMap, dataKey string, log logr.Logger) error
 	odhCertData, ok := cm.Data[dataKey]
 	if !ok || odhCertData == "" {
 		// Print a warning if odh-ca-bundle.crt data is empty
-		return fmt.Errorf("Warning: %s data is empty", dataKey)
+		log.Info(fmt.Sprintf("Warning: %s data is empty", dataKey))
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:370]: Not sure how much sense to log Info message with Warning string :slightly_smiling_face:

**Review by jstourac — COMMENTED** [review]
On lines 301 and 316 of this file, there is following code:

```
		if err := certValidator(cm, "odh-ca-bundle.crt", log); err == nil {
			log.Info("Validating certificates for odh-ca-bundle.crt")
```

Do I understand correctly that by that time the validation had been done already? So shouldn't we rather say something like:

```
Certificates in 'odh-ca-bundle.crt' are valid and can be mounted.
```

Just a proposal.

---

I put a few comments. Basically it can be merged as is, just maybe some of them make sense so the code/logs are slightly clearer.

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 5]
```diff
@@ -309,12 +307,13 @@ func CheckAndMountCACertBundle(ctx context.Context, cli client.Client, notebook
...
 			}
 			volumeMounts = append(volumeMounts, customCAVolumeMounts...)
+			log.Info("Certificates in 'ca-bundle.crt' are valid and can be mounted")
 		} else {
 			log.Error(err, "Error validating certificates for ca-bundle.crt")
 		}
 
 		if err := certValidator(cm, "odh-ca-bundle.crt", log); err == nil {
-			log.Info("Validating certificates for odh-ca-bundle.crt")
+			log.Info("Validating certificates for odh-ca-bundle.crt. They can not be mounted")
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:316]: I don't get this message TBH :thinking: At this line the certificates are validated successfully already, correct?

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 6]
```diff
@@ -324,8 +323,9 @@ func CheckAndMountCACertBundle(ctx context.Context, cli client.Client, notebook
 				},
 			}
 			volumeMounts = append(volumeMounts, customODHCAVolumeMounts...)
+			log.Info("Certificates in 'odh-ca-bundle.crt' are valid and can be mounted")
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:326]: This is something that should be on line 316 instead? I mean isn't it duplicit?

File: components/odh-notebook-controller/controllers/notebook_webhook.go [hunk 7]
```diff
@@ -367,7 +364,8 @@ func certValidator(cm *corev1.ConfigMap, dataKey string, log logr.Logger) error
 	odhCertData, ok := cm.Data[dataKey]
 	if !ok || odhCertData == "" {
 		// Print a warning if odh-ca-bundle.crt data is empty
-		return fmt.Errorf("Warning: %s data is empty", dataKey)
+		log.Info(fmt.Sprintf("%s data are empty", dataKey))
+		return fmt.Errorf("%s data are empty", dataKey)
```
Comment (jstourac) [inline components/odh-notebook-controller/controllers/notebook_webhook.go:368]: Just asking here - so effectively this will lead into the error message in the log as was before (just on line 328), correct? And it is because we actually cannot mount the data as a file with an empty value, is that true?

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 5 above)
Comment (atheo89) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:316]: ohh, you are right! my bad

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 6 above)
Comment (atheo89) [reply components/odh-notebook-controller/controllers/notebook_webhook.go:326]: Hmm, yeah probably is better to keep one Info log for this

File: components/odh-notebook-controller/controllers/notebook_webhook.go (hunk 7 above)
Comment (atheo89) [reply components/odh-notebook-controller/controllers/notebook_webhook.go]: Added a bool variable in case of empty data, to handle it. Now in case of empty data we fall here https://github.com/atheo89/kubeflow/blob/RHOAIENG-4165/components/odh-notebook-controller/controllers/notebook_webhook.go#L322

**Review by jstourac — APPROVED** [review]
Thank you LGTM and if it really behaves like described in the description, I love it :heart_eyes:

**Comment by atheo89** [comment]
> Thank you LGTM and if it really behaves like described in the description, I love it 😍

Yeah, I updated the description as well!

**Comment by jiridanek** [comment]
Does this need any further automated testing?

**Comment by atheo89** [comment]
> Does this need any further automated testing?

Nope, this pr only fixes the displayed logs on the odh-notebook-operator

**Review by harshad16 — APPROVED** [review]
/lgtm
/approve

thank you :100:
